doc/developer/design/20231102_read_holds.md

```diff
@@ -0,0 +1,154 @@
+If difference between the least common physical read frontier of the objects and the `HOLD` time is greater than `ERROR_IF_OLDER_THAN`, the `HOLD`'s read hold is released and the `HOLD` enters an error state.
+(Transactions and `SUBSCRIBE` are not affected, as in the `ALTER HOLD ADVANCE` case.)
+In the error state, `ALTER HOLD ADVANCE` will return an error.
+`SELECT` and `SUBSCRIBE` `AS OF` a time that is between the erroring `HOLD`'s time and the least common physical read frontier will return a helpful error message that the `HOLD` errored because it was not advanced.
+
+It is an error to `DROP` an object that is part of a `HOLD` unless the `DROP` is `CASCADE`.
+If the `DROP` is `CASCADE`, any `HOLD`s on the dropped object are also dropped (even if the `HOLD` is also over other objects).
+
+### Monitoring
+
+To monitor `HOLD`s, a new `mz_catalog.mz_holds` table will be maintained with structure:
+
+- `id text`: the `HOLD`'s id
+- `database_id text`
+- `schema_id text`
+- `name text`
+- `on text[]`: an array of the object ids in the `HOLD`
+- `at uint8`: the `HOLD`'s time
+
+### Implementation of time tracking
+
+The `HOLD` object has an associated time that must be written down somewhere on creation then updated on advancement.
+This will use a persist shard to back it, and is thus similar to a 1-row table.
+As a result of not storing it in the catalog (see Alternatives for the motivation), the `AT <time>` syntax must not be part of the `create_sql` of the `HOLD` object and must be purified away.
+On coordinator bootstrap, a `HOLD` will do a 1-time fetch of its state from the shard to resume its operation.
+
+## Minimal Viable Prototype
+
+An MVP exists at https://github.com/MaterializeInc/materialize/pull/22976.
+
+## Alternatives
+
+- Instead of supporting multiple objects, we could support a single object and enforce the `HOLD` is created in its schema.
+Or, we could support multiple objects, but they must all be in the same schema and the `HOLD` must be in that schema as well.
+
+- These could be called `READ HOLD`s instead of `HOLD`s.
+
+- `SUBSCRIBES` themselves could be made durable, or we introduce `CREATE SUBSCRIPTION` that would similarly achieve this goal ((proposal)[https://github.com/MaterializeInc/materialize/issues/17541#issuecomment-1569340739]).
+
+- Allow users to configure the compaction window of objects.
+This has problems: the window could be too short for recovery or too long and overuse resources, OOMing a cluster.
+Having users routinely advance a `HOLD` should scale the extra resource use to exactly how far behind a user is.
+
+- The stored time could be stored in the catalog instead of persist.
+This would cause a full catalog transaction on any `ALTER HOLD ADVANCE` across the entire customer region, which will eventually not scale (failing our use case isolation goal) since those operations must be serially ordered (even with platform v2).
```

Comment:
Can you speak a little bit more on this. This type of thing feels like the exact thing that belongs in the catalog, metadata about a database object. Do we need to do any type of fencing on this new persist shard?

Once the persist catalog is finished, a catalog transaction will just be a write into persist which sounds equivalent performance wise to updating a the hold metadata persist shard. Though, this might land before the persist catalog.

Reply:
> This type of thing feels like the exact thing that belongs in the catalog, metadata about a database object.

I agree! My gut take was also that this feels better suited to being in the catalog than in a persist shard.

Reply:
The problem I was trying to avoid was a situation where there are hundreds of hold all being advance once per second (assuming once per second if a user is advancing as fast as subscribe is spitting out progress messages). If those are catalog mutations, they need to be serialized. If they only execute one at a time, we have a problem. I think they can go in the catalog as long as we do some kind of write batching so that many read hold advancements can occur in the same catalog txn. I'm sure we could do this, we already do for inserts.

Ok, even if we can, would that amount of catalog traffic (essentially constant transactions) have adverse effects on other DDL statements being run by many different clusters and business units? The catalog itself still has to be serialized, so with hold advancements always happening, maybe all the other DDLs would be slowed down (only a few hundred ms?) waiting for those to commit.

Are these concerns real?

Reply:
With the separate persist shard approach, would there be a single persist shard for all read holds or a persist shard per read hold?

I agree that multiple DDLs a second sounds a bit scary, also implementing group commit for DDL sounds like a lot. Especially with Pv2 work, it makes group commit like things harder to handle. 

You can probably simulate using the catalog by using the config collection. Have the coordinator write something random to the config collection once a second or multiple times a second and see how performance is impacted.

Reply:
How do we make it not scary? I know relatively little about the distributed catalog, but with the current stash based catalog doing a few DDLs per second doesn't sound scary to me. 

I don't think we need general purpose group commit for DDLs—just group commit for `ADVANCE HOLD`. Does that seem less scary?

Reply:
> How do we make it not scary? I know relatively little about the distributed catalog, but with the current stash based catalog doing a few DDLs per second doesn't sound scary to me.

Maybe I'm just being overly nervous then. I just worry about keeping the main Coord thread too busy.

> I don't think we need general purpose group commit for DDLs—just group commit for ADVANCE HOLD. Does that seem less scary?

Yeah I think that's a good idea.